### PR TITLE
[feature] Write data to file on SIGINT unless specified

### PR DESCRIPTION
**Pull request by Fishrock123**
Allows better debugging of applications that have handles that are preventing them from closing.

### CODE REVIEW

File: dprof.js [hunk 1]
```diff
@@ -147,7 +147,17 @@ asyncHook.enable();
 // Print result
 //
 
-process.on('exit', function () {
+if (process.argv.indexOf('--raw-sigint') === -1 ||
```
Comment (AndreasMadsen) [inline dprof.js]: perhaps call it `--dprof-sigint`. The `raw` part doesn't really make sense to me.

**Comment by AndreasMadsen** [comment]
Thanks. You should document this. I guess it is not easy to test with the my odd testrunner, but at the very least you can have the script signal itself.

```js
// one of these
process.argv.push('--raw-sigint');
process.env.DPROF_RAW_SIGINT = '';
require('../../dprof.js');

setInterval(function () {}, 1000):
setTimeout(function () {
  process.kill(process.pid, 'SIGINT');
}, 200);
```

to make the test work:
* add the script file (e.g. `sigint-env.json`) under `scripts`
* add an empty json file (e.g. `sigint-env.json`) under `expected`
* edit `update` to `true` in `test/test.js` (and back when your done).
* check the output manually
* check the output visually using `node test/show.js sigint-env`.

File: dprof.js (hunk 1 above)
Comment (AndreasMadsen) [inline dprof.js]: Also shouldn't this be `!== -1`.

Comment (Fishrock123) [inline dprof.js]: No, the logic is correct -- I think it is better UX to have it as the default, as I suspect that is what most folks will expect.

That's also why I called it `--raw-sigint`, as in, don't touch the default.

It could be something like `--no-sigint-handler` or `--no-dprof-sigint`.

The ENV variable exists for programs that may error on new flags.

Comment (AndreasMadsen) [inline dprof.js]: Should it not be `!process.env.DPROF_RAW_SIGINT` then?

**Comment by Fishrock123** [comment]
@AndreasMadsen updated

File: dprof.js [hunk 2]
```diff
@@ -147,7 +147,17 @@ asyncHook.enable();
 // Print result
 //
 
-process.on('exit', function () {
+if (process.argv.indexOf('--dprof-no-sigint') === -1 ||
+    process.env.DPROF_NO_SIGINT) {
```
Comment (AndreasMadsen) [inline dprof.js]: > No, the logic is correct -- I think it is better UX to have it as the default, as I suspect that is what most folks will expect.

Should it not be `!process.env.DPROF_NO_SIGINT` then?

**Comment by Fishrock123** [comment]
@AndreasMadsen yup, good catch. Fixed I think.

File: dprof.js [hunk 3]
```diff
@@ -147,7 +147,17 @@ asyncHook.enable();
 // Print result
 //
 
-process.on('exit', function () {
+if (process.argv.indexOf('--dprof-no-sigint') === -1 ||
+    !process.env.DPROF_NO_SIGINT) {
+  process.on('SIGINT', () => {
+    writeDataFile();
+    process.exit();
```
Comment (AndreasMadsen) [inline dprof.js]: This indicates a successful exit, however when I SIGINT a process I get the exit code `130`. According to http://unix.stackexchange.com/questions/99112/default-exit-code-when-process-is-terminated exit codes `>= 128` indicates the process exited because of a signal.

I don't know if that is the case here, but surely the exit code should be `>= 1`.

File: test/scripts/sigint-env.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+// one of these
+process.argv.push('--dprof-no-sigint');
```
Comment (AndreasMadsen) [inline test/scripts/sigint-env.js]: Oh. When I originally wrote that comment, it was with the understanding that `SIGINT` handling wasn't the default case. So this will cause `dprof` to not dump anything, in which case the file from the previously executed script that did dump exists.

I think you should just remove these two lines, such it does make a dump and call the file `test/scripts/sigint.js`.

**Comment by Fishrock123** [comment]
Hmmm, the test doesn't seem to fail with either the flag or variable uncommented, not really sure what's up.

**Comment by AndreasMadsen** [comment]
> Hmmm, the test doesn't seem to fail with either the flag or variable uncommented, not really sure what's up.

* `process.on('exit')` is called when `process.exit()` is called from `SIGINT` causing two dumps.
* `process.env.DPROF_NO_SIGINT = ''` is falsy, we need to use `!process.env.hasOwnProperty('DPROF_NO_SIGINT')`
* if either `--dprof-no-sigint` or `DPROF_NO_SIGINT` exists, the if statement shouldn't run. This means that `||` should be `&&`. (De Morgan's laws in formal logic :p).

**Comment by AndreasMadsen** [comment]
I've also updated the test running to be more sensitive to some of these issues and robust in others. There are some conflicts, I think you can just drop 8fc3cb77ee63a84af75cd1f32f06098c3ed4537d and create a new one.

**Comment by Fishrock123** [comment]
> `process.on('exit')` is called when `process.exit()` is called from `SIGINT` causing two dumps.

Arrgh, right. I did some tracing of where the current node behavior comes from, and [found this old issue](https://github.com/nodejs/node-v0.x-archive/issues/457#issuecomment-242872119). Apparently generically calling exit handles is a bad idea. I'm not actually sure how to exit without calling them... `process.abort()` could do but that produces a core file if possible...

**Comment by AndreasMadsen** [comment]
> Arrgh, right. I did some tracing of where the current node behavior comes from, and found this old issue. Apparently generically calling exit handles is a bad idea. I'm not actually sure how to exit without calling them... process.abort() could do but that produces a core file if possible...

"some digging", wow that is very old issue :) A crazy idea is to remove all `SIGINT` handlers and have the process `SIGINT` itself. But I'm also fine with just calling `process.exit()`.

**Comment by Fishrock123** [comment]
Some things:

Note sure why `Signal` says `NaN` for these:

screen shot 2016-08-30 at 11 12 47 am

Also, should we disable `asyncHook` around our own Signal handler so that it doesn't show up?

**Comment by AndreasMadsen** [comment]
> Note sure why Signal says NaN for these.

Thought I fixed that. The `wait` part is because the `destroy` property is `null` (happens when the `destroy` doesn't fire). The solution is to use `root.total` time instead when `destroy: null` (code:  https://github.com/AndreasMadsen/dprof/blob/master/visualizer/info.js#L40). It is not related to this PR so lets fix that separately, I would like this to land.

I don't know why `callback` is `NaN` too. 

> Also, should we disable asyncHook around our own Signal handler so that it doesn't show up?

Was, thinking about that too. If `Signal` handler isn't reused by another `on('SIGINT')` then its safe. But if it is reused then we are removing some information from the user. It may not be a big problem, as the user can pass `--dprof-no-sigint` for full transparency. What do you think?

**Comment by Fishrock123** [comment]
The Signal handling code is at https://github.com/nodejs/node/blob/master/lib/internal/process.js#L180-L219

tl;dr: they are shared.

Maybe we should just go with this for now then?

**Comment by AndreasMadsen** [comment]
> The Signal handling code is at https://github.com/nodejs/node/blob/master/lib/internal/process.js#L180-L219

Thanks for looking that up.

> Maybe we should just go with this for now then?

Definitely. Landed in 2548239572561219281dd79e7fdf2303b152d5cf ee3ee189b54a1db518a1ba8944188314e490992c – Note I rename the test from `sigint-env` to just `sigint`.
